network/peer/peer_test.go

```diff
@@ -273,3 +274,131 @@ func TestSend(t *testing.T) {
 	err = peer1.AwaitClosed(context.Background())
 	require.NoError(err)
 }
+
+func TestPingUptimes(t *testing.T) {
+	trackedSubnetID := ids.GenerateTestID()
+	untrackedSubnetID := ids.GenerateTestID()
+
+	trackedSubnets := set.NewSet[ids.ID](1)
+	trackedSubnets.Add(trackedSubnetID)
+
+	mc := newMessageCreator(t)
+
+	testCases := []struct {
+		name        string
+		msg         message.OutboundMessage
+		shouldClose bool
+		assertFn    func(*require.Assertions, *testPeer)
+	}{
+		{
+			name: "primary network only",
+			msg: func() message.OutboundMessage {
+				pingMsg, err := mc.Ping(1, nil)
+				require.NoError(t, err)
+				return pingMsg
+			}(),
+			assertFn: func(require *require.Assertions, peer *testPeer) {
+				uptime, ok := peer.ObservedUptime(constants.PrimaryNetworkID)
+				require.True(ok)
+				require.Equal(uint32(1), uptime)
+
+				uptime, ok = peer.ObservedUptime(trackedSubnetID)
+				require.False(ok)
+				require.Zero(uptime)
+			},
+		},
+		{
+			name: "primary network and subnet",
+			msg: func() message.OutboundMessage {
+				pingMsg, err := mc.Ping(
+					1,
+					[]*p2p.SubnetUptime{
+						{
+							SubnetId: trackedSubnetID[:],
+							Uptime:   1,
+						},
+					},
+				)
+				require.NoError(t, err)
+				return pingMsg
+			}(),
+			assertFn: func(require *require.Assertions, peer *testPeer) {
+				uptime, ok := peer.ObservedUptime(constants.PrimaryNetworkID)
+				require.True(ok)
+				require.Equal(uint32(1), uptime)
+
+				uptime, ok = peer.ObservedUptime(trackedSubnetID)
+				require.True(ok)
+				require.Equal(uint32(1), uptime)
+			},
+		},
+		{
+			name: "primary network and non tracked subnet",
+			msg: func() message.OutboundMessage {
+				pingMsg, err := mc.Ping(
+					1,
+					[]*p2p.SubnetUptime{
+						{
+							SubnetId: untrackedSubnetID[:],
+							Uptime:   1,
+						},
+						{
+							SubnetId: trackedSubnetID[:],
+							Uptime:   1,
+						},
+					},
+				)
+				require.NoError(t, err)
+				return pingMsg
+			}(),
+			shouldClose: true,
+		},
+	}
+
+	// Note: we reuse peers across tests because makeReadyTestPeers takes awhile
+	// to run.
+	peer0, peer1 := makeReadyTestPeers(t, trackedSubnets)
+	defer func() {
+		peer1.StartClose()
+		peer0.StartClose()
+		require.NoError(t, peer0.AwaitClosed(context.Background()))
+		require.NoError(t, peer1.AwaitClosed(context.Background()))
+	}()
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			require := require.New(t)
+
+			require.True(peer0.Send(context.Background(), tc.msg))
+
+			// Note: shouldClose should only be `true` for the last test because
+			// we reuse peers across tests.
+			if tc.shouldClose {
+				require.NoError(peer1.AwaitClosed(context.Background()))
```

Comment:
why do we simulate or expect the peer to close the connection for the last test? Is it because the network is untracked or is it just the last step in the test?

Reply:
good question - I'll add a comment to this. The peer is closing the connection because the message that it received was invalid.